crates/turborepo-analytics/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+#![deny(clippy::all)]
+
+use std::time::Duration;
+
+use futures::stream::FuturesUnordered;
+use thiserror::Error;
+use tokio::{
+    select,
+    sync::{mpsc, oneshot},
+    task::{JoinError, JoinHandle},
+};
+use tracing::debug;
+use turborepo_api_client::{APIAuth, APIClient};
+pub use turborepo_vercel_api::AnalyticsEvent;
+use uuid::Uuid;
+
+const BUFFER_THRESHOLD: usize = 10;
+
+static EVENT_TIMEOUT: Duration = Duration::from_millis(200);
+static NO_TIMEOUT: Duration = Duration::from_secs(24 * 60 * 60);
+static REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+const CHANNEL_SIZE: usize = 100;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Failed to send analytics event")]
+    SendError(#[from] mpsc::error::SendError<AnalyticsEvent>),
+    #[error("Failed to record analytics")]
+    Join(#[from] JoinError),
+}
+
+// We have two different types because the AnalyticsSender should be shared
+// across threads (i.e. Clone + Send), while the AnalyticsHandle cannot be
+// shared since it contains the structs necessary to shut down the worker.
+pub type AnalyticsSender = mpsc::Sender<AnalyticsEvent>;
+
+pub struct AnalyticsHandle {
+    exit_ch: oneshot::Receiver<()>,
+    handle: JoinHandle<()>,
+}
+
+pub fn start_analytics(api_auth: APIAuth, client: APIClient) -> (AnalyticsSender, AnalyticsHandle) {
+    let (tx, rx) = mpsc::channel(CHANNEL_SIZE);
+    let (cancel_tx, cancel_rx) = oneshot::channel();
+    let session_id = Uuid::new_v4();
+    let worker = Worker {
+        rx,
+        buffer: Vec::new(),
+        session_id,
+        api_auth,
+        senders: FuturesUnordered::new(),
+        exit_ch: cancel_tx,
+        client,
+    };
+    let handle = worker.start();
+
+    let analytics_handle = AnalyticsHandle {
+        exit_ch: cancel_rx,
+        handle,
+    };
+
+    (tx, analytics_handle)
+}
+
+impl AnalyticsHandle {
+    async fn close(self) -> Result<(), Error> {
+        drop(self.exit_ch);
+        self.handle.await?;
+
+        Ok(())
+    }
+
+    pub async fn close_with_timeout(self) {
+        if let Err(err) = tokio::time::timeout(EVENT_TIMEOUT, self.close()).await {
+            debug!("failed to close analytics handle. error: {}", err)
+        }
+    }
```

Comment:
We need a way to send a close signal from a different task via [`ctrl_c`](https://docs.rs/tokio/latest/tokio/signal/fn.ctrl_c.html) so this might not be viable